worker/presence/presence.go

```diff
@@ -0,0 +1,244 @@
+		w.logger.Tracef("request sent")
+	}
+}
+
+func (w *wrapper) forwarderDisconnect(topic string, data forwarder.OriginTarget, err error) {
+	if err != nil {
+		w.logger.Errorf("forwarderChange error %v", err)
+		return
+	}
+	// If we have lost connectivity to the target, we mark the server down.
+	// Ideally this would be when the target is no longer forwarding us messages,
+	// but we aren't guaranteed to get those messages, so we use the lack of our
+	// connectivity to the other machine to indicate that comms are down.
+	if data.Origin == w.origin {
+		w.logger.Tracef("forwarding connection down for %s", data.Target)
+		w.recorder.ServerDown(data.Target)
+	}
+}
+
+func (w *wrapper) agentLogin(topic string, data apiserver.APIConnection, err error) {
+	if err != nil {
+		w.logger.Errorf("agentLogin error %v", err)
+		return
+	}
+	if w.logger.IsTraceEnabled() {
+		agentName := data.AgentTag
+		if data.ControllerAgent {
+			agentName += " (controller)"
+		}
+		w.logger.Tracef("api connect %s:%s -> %s (%v)", data.ModelUUID, agentName, data.Origin, data.ConnectionID)
+	}
+	w.recorder.Connect(data.Origin, data.ModelUUID, data.AgentTag, data.ConnectionID, data.ControllerAgent, data.UserData)
+}
+
+func (w *wrapper) agentDisconnect(topic string, data apiserver.APIConnection, err error) {
+	if err != nil {
+		w.logger.Errorf("agentDisconnect error %v", err)
+		return
+	}
+	w.logger.Tracef("api disconnect %s (%s)", data.Origin, data.ConnectionID)
+	w.recorder.Disconnect(data.Origin, data.ConnectionID)
+}
+
+func (w *wrapper) presenceRequest(topic string, data apiserver.OriginTarget, err error) {
+	if err != nil {
+		w.logger.Errorf("connectionChange error %v", err)
+		return
+	}
+	// If the message is meant for us, respond.
+	if data.Target != w.origin {
+		w.logger.Tracef("presence request for %s ignored, as we are %s", data.Target, w.origin)
+		return
+	}
+
+	w.logger.Tracef("presence request from %s", data.Origin)
+
+	connections := w.recorder.Connections().ForServer(w.origin)
+	values := connections.Values()
+	response := apiserver.PresenceResponse{
+		Connections: make([]apiserver.APIConnection, len(values)),
+	}
+	for i, value := range values {
+		if value.Status != presence.Alive {
+			w.logger.Infof("presence response has weird status: %#v", value)
+		}
+		response.Connections[i] = apiserver.APIConnection{
+			AgentTag:        value.Agent,
+			ControllerAgent: value.ControllerAgent,
+			ModelUUID:       value.Model,
+			ConnectionID:    value.ConnectionID,
+			Origin:          value.Server,
+			UserData:        value.UserData,
+		}
+	}
+	_, err = w.hub.Publish(apiserver.PresenceResponseTopic, response)
+	if err != nil {
+		w.logger.Errorf("cannot send presence response: %v", err)
+	}
+}
+
+func (w *wrapper) presenceResponse(topic string, data apiserver.PresenceResponse, err error) {
+	if err != nil {
+		w.logger.Errorf("connectionChange error %v", err)
+		return
+	}
+	// If this message is from us, ignore it.
+	if data.Origin == w.origin {
```

Comment:
I wonder if a future enhancement could be to do this sort of filtering in the pub sub machinery

Reply:
I agree that we may add something like this.